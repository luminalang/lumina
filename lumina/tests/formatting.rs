//! Snapshot testing to test if formatter changes still look good

use insta;
use lumina_util::test_logger;
use std::path::PathBuf;

fn run(path: &str, file: &str, checked: bool) {
    test_logger();

    if checked {
        let output = lumina::run(path);

        if !output.status.success() {
            panic!("non-success error code: {}", output.status);
        }
    }

    let name = path;

    let manifest = env!("CARGO_MANIFEST_DIR");
    let path = PathBuf::from(format!("{manifest}/../{path}/{file}"));

    let formatted = lumina::formatter::run_file(&path);

    insta::assert_snapshot!(format!("format_{name}"), formatted);
}

#[test]
fn format_hello_world() {
    run("examples/hello-world", "src/main.lm", false);
}

#[test]
fn format_lists() {
    run("examples/lists", "src/main.lm", false);
}

#[test]
fn format_maybe_implementation() {
    run("examples/maybe-implementation", "src/main.lm", false);
}

#[test]
fn format_operators() {
    run("examples/operators", "src/main.lm", false);
}

#[test]
fn format_records() {
    run("examples/records", "src/main.lm", false);
}

#[test]
fn format_tuples() {
    run("examples/tuples", "src/main.lm", false);
}

#[test]
fn format_using_ext_library() {
    run("examples/using-ext-library", "src/main.lm", false);
}

#[test]
fn format_modules() {
    run("examples/modules", "src/main.lm", false);
}

#[cfg(unix)]
#[test]
fn format_ffi() {
    run("examples/ffi", "src/main.lm", false);
}

#[test]
fn format_raw_fn_pointers() {
    run("examples/raw-function-pointers", "src/main.lm", false);
}

#[test]
fn format_fizz_buzz() {
    run("examples/fizz-buzz", "src/main.lm", false);
}
