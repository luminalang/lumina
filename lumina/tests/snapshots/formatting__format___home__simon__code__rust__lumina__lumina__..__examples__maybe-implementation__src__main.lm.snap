---
source: lumina/tests/formatting.rs
expression: formatted
snapshot_kind: text
---
use std:io

// A simple implementation of the `Maybe` sum type
type Maybe a 
  = Just a
  | Nothing
  | Very Long
  | Variant
  | That Keeps Going
  | Hello

fn map f m as (fn a -> b) (Maybe a) -> Maybe b =
  match m
  | Just a -> Just (f a)
  | Nothing -> Nothing

fn and f m as (fn a -> Maybe a) (Maybe a) -> Maybe a =
  match m
  | Just a -> f a
  | Nothing -> Nothing

fn or fallback m as (fn a) (Maybe a) -> a =
  match m
  | Just a -> a
  | Nothing -> fallback

when a can ToString
impl ToString for Maybe a
  fn show m as self -> string =
    match m
    | Nothing -> "Nothing"
    | Just a -> "Just " <> show a

fn main =
  Just 5
  . map #(+ 1)
  . and #(\n -> Just (n + 1))
  . io:println
